src/librustc_trans/trans/type_of.rs

```diff
@@ -405,7 +409,11 @@ pub fn in_memory_type_of<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>, t: Ty<'tcx>) ->
       }
       ty::TyStruct(did, ref substs) => {
           if t.is_simd(cx.tcx()) {
-              let llet = in_memory_type_of(cx, t.simd_type(cx.tcx()));
+              let elem_ty = match t.simd_machine_type(cx.tcx()) {
+                  Some(e) => e,
+                  None => cx.sess().fatal("monomorphising SIMD type to an incompatible element")
```

Comment:
This is ugly... is allowing generic SIMD types actually important in practice?  It seems like you could accomplish most of what is necessary with traits and maybe a few macros.